src/haswell/implementation.cpp

```diff
@@ -117,6 +153,47 @@ simdutf_warn_unused size_t implementation::utf16_length_from_utf8(const char * i
   return utf8::utf16_length_from_utf8(input, length);
 }
 
+simdutf_warn_unused size_t implementation::utf8_length_from_utf32(const char32_t * input, size_t length) const noexcept {
+  const __m256i v_00000000 = _mm256_setzero_si256();
+  const __m256i v_ffffff80 = _mm256_set1_epi32((int32_t)0xffffff80);
+  const __m256i v_fffff800 = _mm256_set1_epi32((int32_t)0xfffff800);
+  const __m256i v_ffff0000 = _mm256_set1_epi32((int32_t)0xffff0000);
+  size_t pos = 0;
+  size_t count = 0;
+  for(;pos + 8 <= length; pos += 8) {
+    __m256i in = _mm256_loadu_si256((__m256i*)(input + pos));
+    const __m256i ascii_bytes_bytemask = _mm256_cmpeq_epi32(_mm256_and_si256(in, v_ffffff80), v_00000000);
+    const __m256i one_two_bytes_bytemask = _mm256_cmpeq_epi32(_mm256_and_si256(in, v_fffff800), v_00000000);
+    const __m256i two_bytes_bytemask = _mm256_xor_si256(one_two_bytes_bytemask, ascii_bytes_bytemask);
+    const __m256i one_two_three_bytes_bytemask = _mm256_cmpeq_epi32(_mm256_and_si256(in, v_ffff0000), v_00000000);
+    const __m256i three_bytes_bytemask = _mm256_xor_si256(one_two_three_bytes_bytemask, one_two_bytes_bytemask);
+    const uint32_t ascii_bytes_bitmask = static_cast<uint32_t>(_mm256_movemask_epi8(ascii_bytes_bytemask));
+    const uint32_t two_bytes_bitmask = static_cast<uint32_t>(_mm256_movemask_epi8(two_bytes_bytemask));
+    const uint32_t three_bytes_bitmask = static_cast<uint32_t>(_mm256_movemask_epi8(three_bytes_bytemask));
+
+    size_t ascii_count = count_ones(ascii_bytes_bitmask) / 4;
+    size_t two_bytes_count = count_ones(two_bytes_bitmask) / 4;
+    size_t three_bytes_count = count_ones(three_bytes_bitmask) / 4;
+    count += 32 - 3*ascii_count - 2*two_bytes_count - three_bytes_count;
+  }
+  return count + scalar::utf32::utf8_length_from_utf32(input + pos, length - pos);
+}
+
+simdutf_warn_unused size_t implementation::utf16_length_from_utf32(const char32_t * input, size_t length) const noexcept {
+  const __m256i v_00000000 = _mm256_setzero_si256();
+  const __m256i v_ffff0000 = _mm256_set1_epi32((int32_t)0xffff0000);
+  size_t pos = 0;
+  size_t count = 0;
+  for(;pos + 8 <= length; pos += 8) {
+    __m256i in = _mm256_loadu_si256((__m256i*)(input + pos));
+    const __m256i surrogate_bytemask = _mm256_cmpeq_epi32(_mm256_and_si256(in, v_ffff0000), v_00000000);
+    const uint32_t surrogate_bitmask = static_cast<uint32_t>(_mm256_movemask_epi8(surrogate_bytemask));
+    size_t surrogate_count = count_ones(~surrogate_bitmask)/4;
```

Comment:
That's potentially confusing code. The count_ones functions takes in an uint64_t value. You take surrogate_bitmask which you negate and pass it to count_ones. 
Maybe the following would be less error-prone: (32-count_ones(surrogate_bitmask))/4.